### PR TITLE
Adding further engine bindings

### PR DESCRIPTION
**Pull request by whytro**
Added more bindings for the other engine calls. However, I did run into an issue with the MatchParameters bindings, which I suspect has something to do with the way the argument forwarding is set up.

Also added `__repr__` outputs for the custom JSON types, so they print out as expected.

### CODE REVIEW

**Comment by whytro** [comment]
I've attached the MatchParameters file as reference [here](https://github.com/gis-ops/py-osrm/pull/6/files#diff-9619bc7b400b11861b65218c112b03e137e032454402a404831a71779d0e314d), but the gist of it is that I can't find the proper parameter combination to have it link. It seems to not like that waypoints is explicitly specified in MatchParameters, while it's also handled in RouteParameters, so it expects the waypoints parameters again, which also throws off the va starting point for BaseParameters.

File: include/types/jsoncontainer_nb.h [hunk 1]
```diff
@@ -1,8 +1,115 @@
...
+    static handle from_cpp(const json::Null&, rv_policy, cleanup_list*) noexcept {
+        return none().release();
+    }
+};
+
+} //nanobind::detail
+
+struct ValueStringifyVisitor {
+    ValueStringifyVisitor(std::string& output) : _output(output) {};
+    std::string& _output;
```
Comment (nilsnolde) [inline include/types/jsoncontainer_nb.h:65]: seems a bit more complicated than necessary? couldn't we just return the string value from each operator? unless it's needed to concatenate multiple values from a single `ValueStringifyVisitor` instance?

File: include/types/jsoncontainer_nb.h [hunk 2]
```diff
@@ -1,8 +1,115 @@
...
+                _output += ", ";
+            }
+            _output += "'" + itr.first + "': ";
+            mapbox::util::apply_visitor(*this, itr.second);
+            first = false;
+        }
+        _output += "}";
+    }
+    void operator()(const mapbox::util::recursive_wrapper<json::Object>& obj) {
+        visitobject(obj);
```
Comment (nilsnolde) [inline include/types/jsoncontainer_nb.h:111]: does this need a `obj.get()`?

**Comment by nilsnolde** [comment]
Thanks for the further work! It's quite a lot to review though;) It's ok for now, and let's merge once reviewed and got your small problem out of the way. After that, we should concentrate on testing IMO, which will also allow us to see how the API currently works for a user (it's hard to understand just looking at the code).

File: include/types/jsoncontainer_nb.h [hunk 3]
```diff
@@ -1,8 +1,115 @@
...
+    }
+};
+
+} //nanobind::detail
+
+struct ValueStringifyVisitor {
+    ValueStringifyVisitor(std::string& output) : _output(output) {};
+    std::string& _output;
+
+    void operator()(const json::String& str) {
```
Comment (SiarheiFedartsou) [inline include/types/jsoncontainer_nb.h:67]:
```suggestion
    std::string operator()(const json::String& str) {
```

File: src/parameters/tileparameter_nb.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    using osrm::engine::api::TileParameters;
+
+    nb::class_<TileParameters>(m, "TileParameters", nb::is_final())
+        .def(nb::init<>())
+        .def(nb::init<unsigned, unsigned, unsigned>(),
+            "x"_a, "y"_a, "z"_a
+        )
+        .def("__init__", [](TileParameters* t, const std::vector<unsigned>& coord) {
+            new (t) TileParameters{coord[0], coord[1], coord[2]};
+         })
```
Comment (nilsnolde) [inline src/parameters/tileparameter_nb.cpp:22]: just for reference: better to check in the function body if it was exactly 3 parameters passed, no more/less. also the result here is very likely binary protobuf

Comment (whytro) [reply src/parameters/tileparameter_nb.cpp]: Okay, I've adjusted this to validate the vector, and throw a runtime error if the size is not 3.

File: include/types/jsoncontainer_nb.h (hunk 3 above)
Comment (whytro) [reply include/types/jsoncontainer_nb.h]: I've adjusted the visitor to return string values accordingly!

File: src/parameters/matchparameter_nb.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                    std::vector<osrm::util::Coordinate> coordinates,
+                    std::vector<boost::optional<osrm::engine::Hint>> hints,
+                    std::vector<boost::optional<double>> radiuses,
+                    std::vector<boost::optional<osrm::engine::Bearing>> bearings,
+                    std::vector<boost::optional<osrm::engine::Approach>> approaches,
+                    bool generate_hints,
+                    std::vector<std::string> exclude,
+                    const BaseParameters::SnappingType snapping
+            ) {
+                new (t) MatchParameters(
```
Comment (whytro) [inline src/parameters/matchparameter_nb.cpp:44]: I've done the MatchParameters in the "assigning" way, but I'm not sure if it would be clearer to just fully assign all values manually, or use the constructor for MatchParameters and RouteParameters, and then manually assign for BaseParameters. 

I believe the NodeJS bindings do it in the inverse way instead (ie. assigning BaseParameters, and then everything else is explicitly assigned), but I think that would be more complicated to replicate here.

Comment (nilsnolde) [reply src/parameters/matchparameter_nb.cpp:44]: > but I'm not sure if it would be clearer to just fully assign all values manually

Yeah I agree here

File: include/types/jsoncontainer_nb.h [hunk 6]
```diff
@@ -1,8 +1,112 @@
...
+            if(i != 0) {
+                output += ", ";
+            }
+            output += mapbox::util::apply_visitor(*this, arr.values[i]);
+        }
+        return output + "]";
+    }
+    std::string operator()(const mapbox::util::recursive_wrapper<json::Array>& arr) {
+        return visitarray(arr.get());
+    }
```
Comment (nilsnolde) [inline include/types/jsoncontainer_nb.h:92]: and this still works for recursive arrays?

**Review by nilsnolde — COMMENTED** [review]
Thanks for the updates. Just a question for the JSON return type commit, then it LGTM. Sorry, if we were confusing on that, I'm just not sure your previous solution was actually working better for recursive types such as array/object.

**Comment by whytro** [comment]
> Thanks for the updates. Just a question for the JSON return type commit, then it LGTM. Sorry, if we were confusing on that, I'm just not sure your previous solution was actually working better for recursive types such as array/object.

Yes, it works both ways, and basic benchmarking seemed to show negligible difference between the two methods, so I went with the return visitor for code clarity.

On the topic of mapbox recursive types, it seems it works without an explicit `.get()` call, but I added it in anyways for clarity.

File: src/parameters/matchparameter_nb.cpp (hunk 5 above)
Comment (nilsnolde) [reply src/parameters/matchparameter_nb.cpp:44]: > you agree with manually assigning all values

exactly that:)

Comment (nilsnolde) [reply src/parameters/matchparameter_nb.cpp:44]: you want to integrate that change before merging or make an issue out of it?

Comment (whytro) [reply src/parameters/matchparameter_nb.cpp:44]: After taking a closer look, it seems like default initializing MatchParameters and setting each parameter explicitly might result in additional complication. This is because when values are passed into the parent constructor, the parent constructor does perform some more transformations and assignments of its own.

For example, `AnnotationsType` gets converted to a boolean value, and the `number_of_alternatives` value is also calculated based off the passed `alternatives` value. As a result, I would have to copy over those calculations to the bindings side (ie. `number_of_alternatives = alternatives ? 1u : 0u`).

As a result, I think this approach may lose code clarity, since those value calculations would have to be replicated on the bindings side, and also require additional copied hardcoded values/calculations.

Comment (SiarheiFedartsou) [reply src/parameters/matchparameter_nb.cpp:44]: Btw boolean `alternatives` is deprecated as well https://github.com/Project-OSRM/osrm-backend/blob/62e11fd5dbc8d5e888c6e2555505ae684d8a2759/include/engine/api/route_parameters.hpp#L168

Comment (whytro) [reply src/parameters/matchparameter_nb.cpp:44]: Ah, if it's `alternatives` bool input is getting deprecated, should we still support bool input for the Python bindings? 

The part that I didn't fully understand about the NodeJS `annotations_type` part there is that it seems to check what enum type was sent for `annotations` to assign the `annotations_type` value (in an array), but for the C++ side of things, it seems to base it off which RouteParameters constructor is called. If I'm understanding this part correctly, the user is able to send in an array of AnnotationsTypes, which gets calculated up? In that case, it would make sense to apply that same abstraction to the Python bindings.

Comment (nilsnolde) [reply src/parameters/matchparameter_nb.cpp:44]: > Ah, if its alternatives bool input is getting deprecated, should we still support bool input for the Python bindings?

IMO no. Boolean never really made sense there and it should be int. 

The other stuff @SiarheiFedartsou is better suited answering:)

Comment (SiarheiFedartsou) [reply src/parameters/matchparameter_nb.cpp:44]: > The part that I didn't fully understand about the NodeJS annotations_type part there is that it seems to check what enum type was sent for annotations to assign the annotations_type value (in an array), but for the C++ side of things, it seems to base it off which RouteParameters constructor is called. If I'm understanding this part correctly, the user is able to send in an array of AnnotationsTypes, which gets calculated up? In that case, it would make sense to apply that same abstraction to the Python bindings.

Not sure I fully understand the question... What happens in NodeJS bindings it simply converts something like this on JS side `["nodes", "distance", "speed"]` to smth like this on C++ side `std::vector<AnnotationType>{AnnotationType::Nodes, AnnotationType::Distance, AnnotationType::Speed}`. Can we do the same on Python side? Ofc we can use https://docs.python.org/3/library/enum.html or smth like this if needed.

Comment (SiarheiFedartsou) [reply src/parameters/matchparameter_nb.cpp:44]: ahhh, I got it. `annotations_type` is not vector indeed, it is [bitmask](https://en.wikipedia.org/wiki/Mask_(computing)). Then you indeed can follow NodeJS path and just transform Python's list to this bitmask just like NodeJS bindings do. 

 So I would rephrase what I said above:
 
 > Not sure I fully understand the question... What happens in NodeJS bindings it simply converts something like this on JS side `["nodes", "distance", "speed"]` to smth like this on C++ side `AnnotationType::Nodes | AnnotationType::Distance | AnnotationType::Speed`. Can we do the same on Python side? Ofc we can use https://docs.python.org/3/library/enum.html or smth like this if needed.

Comment (whytro) [reply src/parameters/matchparameter_nb.cpp:44]: Okay, that sounds doable, thanks for the clarifications!

Comment (whytro) [reply src/parameters/matchparameter_nb.cpp]: I've made adjustments to accommodate the requested changes. In the process, I've also discovered a bug in osrm-backend's RouteParameter's AnnotationsType `|=` operator overload that renders it nonfunctional, so I've also put in a [PR over there](https://github.com/Project-OSRM/osrm-backend/pull/6646) to address that.

**Comment by nilsnolde** [comment]
LGTM, thanks!
